README.md
# Library and codes for nodes implementing LoRaWAN Class C #

Please refer to Library_README.md for the library usage

The library is based-off work from : https://github.com/antaresdocumentation/lorawan-loraid \
I modified bits and pieces in the library to suit Malaysia AS1 frequencies as well as some minor changes at some of the places.

I ran the code on Arduino UNO + Dragino LoRa Shield + Relay Shield V3.0.

Dragino LoRa Shield : https://wzper.my/index.php?id_product=90&controller=product \
Relay Shield V3.0 : https://my.cytron.io/p-relay-shield-v3.0?r=1 

As i am interested to work on Class C in order to control actuator or Relay using LoRaWAN, we will focus only on Class C here. \
For normal Class A, you can refer to my other repository : https://github.com/zakibakar75/arduino-lmic-as923.

This library is not based on lmic by the way. It was originally written by Ideetron BV. It works pretty good.
The only feature is missing is ADR functionality. But it is just fine to make your node works.

As TTN is currently does not support Class C yet, i am using Gotthardp LoRaWAN Server for my work - https://github.com/gotthardp/lorawan-server \
I installed the lorawan server on my Raspberry Pi 3B+. It works pretty good.

In my setup, i am using Dragino LG308 Multi-Channel Gateway which talks to the lorawan server to get the data processed. 

LoRaWAN Gateway : https://wzper.my/index.php?id_product=105&controller=product

In the same Raspberry Pi, i also installed Mosquitto MQTT broker.
Then, i connect the lorawan server with the MQTT broker at the "connector" section of the lorawan server.

To receive and send data from/to the node, i am using Node-Red on my personal computer. \
Incoming data from node can be seen on Node-Red. \
We can also control the Relay from within the Node-Red flow.

Some settings on the Lorawan server must be done in order to make things work perfectly. \
Here, you can find the screenshot of my setting. \
Also, the flow for my Node-Red for your convenience.


doc/Release_Notes.txt
Release Notes to be put here.


src/arduino-rfm/Commands.cpp
/******************************************************************************************
* Copyright 2017 Ideetron B.V.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
******************************************************************************************/
/****************************************************************************************
* File:     Commands.cpp
* Author:   Gerben den Hartog
* Compagny: Ideetron B.V.
* Website:  http://www.ideetron.nl/LoRa
* E-mail:   info@ideetron.nl
****************************************************************************************/
/****************************************************************************************
* Created on:         12-01-2017
* Supported Hardware: ID150119-02 Nexus board with RFM95
****************************************************************************************/

/*
*****************************************************************************************
* INCLUDE FILES
*****************************************************************************************
*/

#include <SPI.h>
#include "Commands.h"
#include "Conversions.h"
#include "RFM95.h"
#include "Struct.h"
#include "Config.h"

void UART_Send_Newline()
{
  Serial.write(0x0D); //Carriage return
  Serial.write(0x0A); //Line feed
}

void UART_Send_Data(unsigned char *Data, unsigned char Length)
{
  unsigned char i;
  unsigned char Upper_Nibble;
  unsigned char Lower_Nibble;

  for(i = 0; i < Length; i++)
  {
    Hex2ASCII(Data[i], &Upper_Nibble, &Lower_Nibble);

    //Send the data
    Serial.write(Upper_Nibble);
    Serial.write(Lower_Nibble);
  }
}

void UART_Send_Datarate(unsigned char *Datarate)
{

  switch(*Datarate)
  {
      case 0x00:
          Serial.write("SF 12 BW 125");
          break;
      case 0x01:
          Serial.write("SF 11 BW 125");
          break;
      case 0x02:
          Serial.write("SF 10 BW 125");
          break;
      case 0x03:
          Serial.write("SF 9 BW 125");
          break;
      case 0x04:
          Serial.write("SF 8 BW 125");
          break;
      case 0x05:
          Serial.write("SF 7 BW 125");
          break;
      case 0x06:
          Serial.write("SF 7 BW 250");
          break;
  }

  UART_Send_Newline();
}

void UART_Send_Channel(unsigned char *Channel)
{
  switch(*Channel)
  {
    #ifdef AS_923
    case 0x00:
      Serial.write("923.200");
      break;
    case 0x01:
      Serial.write("923.400");
      break;
    case 0x02:
      Serial.write("922.200");
      break;
    case 0x03:
      Serial.write("922.400");
      break;
    case 0x04:
      Serial.write("922.600");
      break;
    case 0x05:
      Serial.write("922.800");
      break;
    case 0x06:
      Serial.write("923.000");
      break;
    case 0x07:
      Serial.write("922.000");
      break;
    case 0x08:
      Serial.write("924.500");
      break;
    case 0x10:
      Serial.write("923.200");
      break;
    #else
    case 0x00:
      Serial.write("868.100");
      break;
    case 0x01:
      Serial.write("868.300");
      break;
    case 0x02:
      Serial.write("868.500");
      break;
    case 0x03:
      Serial.write("867.100");
      break;
    case 0x04:
      Serial.write("867.300");
      break;
    case 0x05:
      Serial.write("867.500");
      break;
    case 0x06:
      Serial.write("867.700");
      break;
    case 0x07:
      Serial.write("867.900");
      break;
    case 0x10:
      Serial.write("869.525");
      break;
    #endif
  }

    UART_Send_Newline();

}

void Mac_DevAddr(unsigned char *buffer, unsigned char *DevAddr)
{

  //Check if it is a set command and there is enough data sent
  DevAddr[0] = ASCII2Hex(buffer[0],buffer[1]);
  DevAddr[1] = ASCII2Hex(buffer[2],buffer[3]);
  DevAddr[2] = ASCII2Hex(buffer[4],buffer[5]);
  DevAddr[3] = ASCII2Hex(buffer[6],buffer[7]);

  //Send set DevAddr
  Serial.write("DevAddr: ");
  UART_Send_Data(DevAddr, 0x04);
  UART_Send_Newline();
  UART_Send_Newline();
}

void Mac_NwkSKey(unsigned char *buffer, unsigned char *NwkSKey)
{
  unsigned char i;

  for(i = 0; i < 16; i++)
  {
    NwkSKey[i] = ASCII2Hex(buffer[i*2],buffer[(i*2)+1]);
  }

  //Send NwkSkey
  Serial.write("NwkSKey: ");
  UART_Send_Data(NwkSKey,0x10);
  UART_Send_Newline();
}

void Mac_AppSKey(unsigned char *buffer, unsigned char *AppSKey)
{
  unsigned char i;

  for(i = 0; i < 16; i++)
  {
    AppSKey[i] = ASCII2Hex(buffer[i*2],buffer[(i*2)+1]);
  }

  //Send AppSkey
  Serial.write("AppSKey: ");
  UART_Send_Data(AppSKey,0x10);
  UART_Send_Newline();
}

void Mac_AppKey(unsigned char *buffer, unsigned char *AppKey)
{
  unsigned char i;

  for(i = 0; i < 16; i++)
  {
    AppKey[i] = ASCII2Hex(buffer[i*2],buffer[(i*2)+1]);
  }

  //Send AppKey
  Serial.write("AppKey: ");
  UART_Send_Data(AppKey,0x10);
  UART_Send_Newline();

}

void Mac_AppEUI(unsigned char *buffer, unsigned char *AppEUI)
{
  unsigned char i;

  for(i = 0; i < 8; i++)
  {
    AppEUI[i] = ASCII2Hex(buffer[i*2],buffer[(i*2)+1]);
  }

  //Send AppEUI
  Serial.write("AppEUI: ");
  UART_Send_Data(AppEUI,0x08);
  UART_Send_Newline();
}

void Mac_DevEUI(unsigned char *buffer, unsigned char *DevEUI)
{
  unsigned char i;

  for(i = 0; i < 8; i++)
  {
    DevEUI[i] = ASCII2Hex(buffer[i*2],buffer[(i*2)+1]);
  }

  //Send DevEUI
  Serial.write("DevEUI: ");
  UART_Send_Data(DevEUI,0x08);
  UART_Send_Newline();
}

void Mac_DrTx(unsigned char drate_tx, unsigned char *Datarate)
{
  unsigned char Datarate_Temp;

  //Convert to temp
  Datarate_Temp = drate_tx;

  //Check if the value is oke
  if(Datarate_Temp <= 0x06)
  {
    *Datarate = Datarate_Temp;
  }

  Serial.write("Datarate Tx: ");

  UART_Send_Datarate(Datarate);
}

void Mac_DrRx(unsigned char drate_rx, unsigned char *Datarate)
{
  unsigned char Datarate_Temp;

  //Convert to temp
  Datarate_Temp = drate_rx;

  //Check if the value is oke
  if(Datarate_Temp <= 0x06)
  {
    *Datarate = Datarate_Temp;
  }

  Serial.write("Datarate Rx: ");

  UART_Send_Datarate(Datarate);
}

void Mac_ChTx(unsigned char ch_tx_idx, unsigned char *Channel)
{
 unsigned char Channel_Temp;

  //Convert to temp
  Channel_Temp = ch_tx_idx;

  //Check if the value is oke
  #ifdef AS_923
  if(Channel_Temp <= 0x08 || Channel_Temp == 0x10)
  {
    *Channel = Channel_Temp;
  }
  #else
  if(Channel_Temp <= 0x07 || Channel_Temp == 0x10)
  {
    *Channel = Channel_Temp;
  }
  #endif

  Serial.write("Channel Tx: ");

  UART_Send_Channel(Channel);
}

void Mac_ChRx(unsigned char ch_rx_idx, unsigned char *Channel)
{
  unsigned char Channel_Temp;

  //Convert to temp
  Channel_Temp = ch_rx_idx;

  //Check if the value is oke
  #ifdef AS_923
  if(Channel_Temp <= 0x08 || Channel_Temp == 0x10)
  {
    *Channel = Channel_Temp;
  }
  #else
  if(Channel_Temp <= 0x07 || Channel_Temp == 0x10)
  {
    *Channel = Channel_Temp;
  }
  #endif

  Serial.write("Channel Rx: ");

  UART_Send_Channel(Channel);
}

void Mac_Power(unsigned char pwr_idx, unsigned char *Power)
{

  unsigned char RFM_Data;

  *Power = pwr_idx;

  //Check if power is not over 0x0F
  if(*Power > 0x0F)
  {
    *Power = 0x0F;
  }

  //Set all ther correct bits for the RFM register
  RFM_Data = *Power + 0xF0;

  //Write power to RFM module
  RFM_Write(0x09,RFM_Data);

  //Send answer
  Serial.write("Power: ");
  UART_Send_Data(Power,0x01);
  UART_Send_Newline();
}

void Mac_Confirm(unsigned char confirm_msg, unsigned char *Confirm)
{

  *Confirm = confirm_msg;
  if(*Confirm >= 0x01)
  {
    *Confirm = 0x01;
  }

  //Send answer
  Serial.write("Confirm: ");
  UART_Send_Data(Confirm,0x01);
  UART_Send_Newline();
}

void Mac_Channel_Hopping(unsigned char hop_enable, unsigned char *Channel_Hopping)
{
  *Channel_Hopping = hop_enable;

  if(*Channel_Hopping >= 0x01)
  {
    *Channel_Hopping = 0x01;
  }

  //Send answer
  Serial.write("Channel Hopping: ");
  UART_Send_Data(Channel_Hopping,0x01);
  UART_Send_Newline();
}

void Mac_Class(devclass_t dev_class, sSettings *LoRa_Settings)
{
  if(dev_class == CLASS_A)
  {
    LoRa_Settings->Mote_Class = 0x00;
  }
  else
  {
    LoRa_Settings->Mote_Class = 0x01;
  }

  //Send answer and switch rfm to standby or receive
  Serial.write("Mote Class: ");
  if(LoRa_Settings->Mote_Class == 0x00)
  {

    //Switch RFM to standby
    RFM_Switch_Mode(0x01);

    Serial.write("A");
  }
  else
  {
    //Switch RFM to continuou receive
    RFM_Continuous_Receive(LoRa_Settings);

    Serial.write("C");
  }
  UART_Send_Newline();
}

void Mac_Data(unsigned char *buffer, unsigned int len, sBuffer *RFM_Buffer)
{
  RFM_Buffer->Counter = len;
  memcpy(RFM_Buffer->Data, buffer, len);
  // unsigned int len_temp = len;
  // unsigned char i;

  // //Check for an even number of received data bytes other wise fill whit 0x00
  // if(len % 2)
  // {
  //   buffer[len] = 0x00;
  //   len_temp++;
  // }

  // RFM_Buffer->Counter = len_temp / 2;

  // for(i = 0x00; i < RFM_Buffer->Counter; i++)
  // {
  //   RFM_Buffer->Data[i] = ASCII2Hex(buffer[i*2],buffer[(i*2)+1]);
  // }

  Serial.write("Data : ");
  UART_Send_Data(RFM_Buffer->Data,RFM_Buffer->Counter);
  UART_Send_Newline();
}


src/arduino-rfm/LoRaMAC.cpp
/******************************************************************************************
* Copyright 2017 Ideetron B.V.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Lesser General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Lesser General Public License for more details.
*
* You should have received a copy of the GNU Lesser General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
******************************************************************************************/
/****************************************************************************************
* File:     LoRaMAC.cpp
* Author:   Gerben den Hartog
* Compagny: Ideetron B.V.
* Website:  http://www.ideetron.nl/LoRa
* E-mail:   info@ideetron.nl
****************************************************************************************/
/****************************************************************************************
* Created on:         06-01-2017
* Supported Hardware: ID150119-02 Nexus board with RFM95
****************************************************************************************/
/*
*****************************************************************************************
* INCLUDE FILES
*****************************************************************************************
*/

#include <SPI.h>
#include "AES-128.h"
#include "RFM95.h"
#include "Encrypt.h"
#include "LoRaMAC.h"
#include "Struct.h"
#include "Commands.h"
#include "loraid-arduino-rfm.h"

/*
*****************************************************************************************
* FUNCTIONS
*****************************************************************************************
*/


/*
*****************************************************************************************
* Description : Function that handles a send and receive cycle with timing for receive slots.
*				This function is only used for Class A motes. The wait times are tested with
*				the iot.semtech.com site.
*
* Arguments   : *Data_Tx pointer to tranmit buffer
*				*Data_Rx pointer to receive buffer
*				*RFM_Command pointer to current RFM state
*				*Session_Data pointer to sLoRa_Session sturct
*				*OTAA_Data pointer to sLoRa_OTAA struct
*				*Message_Rx pointer to sLoRa_Message struct used for the received message information
*				*LoRa_Settings pointer to sSetting struct
*****************************************************************************************
*/
void LORA_Cycle(sBuffer *Data_Tx, sBuffer *Data_Rx, RFM_command_t *RFM_Command, sLoRa_Session *Session_Data, sLoRa_OTAA *OTAA_Data, sLoRa_Message *Message_Rx, sSettings *LoRa_Settings)
{
	unsigned char i;
	unsigned int Receive_Delay_1 = 950;
	unsigned int Receive_Delay_2 = 1950;
	unsigned int Receive_Delay_JoinAck = 5950;
	unsigned long prevTime = 0;

	unsigned char Channel_Rx_1 = LoRa_Settings->Channel_Tx;
	unsigned char Datarate_Rx_1 = LoRa_Settings->Datarate_Tx;
	unsigned char Channel_Rx_2 = LoRa_Settings->Channel_Rx;
	unsigned char Datarate_Rx_2 = LoRa_Settings->Datarate_Rx;

	if(*RFM_Command == JOIN)
  	{
  	  //Send join Reqest message
  	  LoRa_Send_JoinReq(OTAA_Data, LoRa_Settings);

      delay(Receive_Delay_JoinAck);
  	}

	//Send normal data message
  	if(*RFM_Command == NEW_RFM_COMMAND)
  	{
  	  LORA_Send_Data(Data_Tx, Session_Data, LoRa_Settings);

	  // Start timer
	  prevTime = millis();

	  // Class C open RX2 immediately after sending data
      if(LoRa_Settings->Mote_Class == 0x01)
	  {
	    // RX2 window
        LoRa_Settings->Channel_Rx = Channel_Rx_2;    // set Rx2 channel
        LoRa_Settings->Datarate_Rx = Datarate_Rx_2;   //set data rate Rx2
	  	LORA_Receive_Data(Data_Rx, Session_Data, OTAA_Data, Message_Rx, LoRa_Settings);
		if(Data_Rx->Counter > 0) {
			Serial.print((char *)Data_Rx->Data);
		} else {
			Serial.println("No Data RX2 Class C");
		}
	  }

	  // Wait for RX1 delay
	  while(millis() - prevTime < Receive_Delay_1);

	  // Add RX1 window
      LoRa_Settings->Channel_Rx = Channel_Rx_1;    // set Rx1 channel, same with Tx
      LoRa_Settings->Datarate_Rx = Datarate_Rx_1;   //set data rate Rx1, same with Tx

	  LORA_Receive_Data(Data_Rx, Session_Data, OTAA_Data, Message_Rx, LoRa_Settings);
	  if(Data_Rx->Counter > 0) {
		Serial.print((char *)Data_Rx->Data);
	  } else {
		Serial.println("No Data RX1");
	  }
      // Wait for RX2 delay
      while(millis() - prevTime < Receive_Delay_2);

	  // RX2 window
      LoRa_Settings->Channel_Rx = Channel_Rx_2;    // set Rx2 channel
      LoRa_Settings->Datarate_Rx = Datarate_Rx_2;   //set data rate Rx2
    }

    LORA_Receive_Data(Data_Rx, Session_Data, OTAA_Data, Message_Rx, LoRa_Settings);
	if(Data_Rx->Counter > 0) {
		Serial.print((char *)Data_Rx->Data);
	} else {
		Serial.println("No Data RX2");
	}
}

/*
*****************************************************************************************
* Description : Function that is used to build a LoRaWAN data message and then tranmit it.
*
* Arguments   : *Data_Tx pointer to tranmit buffer
*				*Session_Data pointer to sLoRa_Session sturct
*				*LoRa_Settings pointer to sSetting struct
*****************************************************************************************
*/
void LORA_Send_Data(sBuffer *Data_Tx, sLoRa_Session *Session_Data, sSettings *LoRa_Settings)
{
  //Define variables
  unsigned char i;

  //Initialise RFM buffer
  unsigned char RFM_Data[64];
  sBuffer RFM_Package = {&RFM_Data[0], 0x00};

  //Initialise Message struct for a transmit message
  sLoRa_Message Message;

  Message.MAC_Header = 0x00;
  Message.Frame_Port = 0x01; //Frame port always 1 for now
  Message.Frame_Control = 0x00;

  //Load device addres from session data into the message
  Message.DevAddr[0] = Session_Data->DevAddr[0];
  Message.DevAddr[1] = Session_Data->DevAddr[1];
  Message.DevAddr[2] = Session_Data->DevAddr[2];
  Message.DevAddr[3] = Session_Data->DevAddr[3];

  //Set up direction
  Message.Direction = 0x00;

  //Load the frame counter from the session data into the message
  Message.Frame_Counter = *Session_Data->Frame_Counter;

  //Set confirmation
  //Unconfirmed
  if(LoRa_Settings->Confirm == 0x00)
  {
      Message.MAC_Header = Message.MAC_Header | 0x40;
  }
  //Confirmed
  else
  {
      Message.MAC_Header = Message.MAC_Header | 0x80;
  }

  //Build the Radio Package
  //Load mac header
  RFM_Data[0] = Message.MAC_Header;

  //Load device address
  RFM_Data[1] = Message.DevAddr[3];
  RFM_Data[2] = Message.DevAddr[2];
  RFM_Data[3] = Message.DevAddr[1];
  RFM_Data[4] = Message.DevAddr[0];

  //Load frame control
  RFM_Data[5] = Message.Frame_Control;

  //Load frame counter
  RFM_Data[6] = (*Session_Data->Frame_Counter & 0x00FF);
  RFM_Data[7] = ((*Session_Data->Frame_Counter >> 8) & 0x00FF);

  //Set data counter to 8
  RFM_Package.Counter = 8;

  //If there is data load the Frame_Port field
  //Encrypt the data and load the data
  if(Data_Tx->Counter > 0x00)
  {
	//Load Frame port field
    RFM_Data[8] = Message.Frame_Port;

    //Raise package counter
    RFM_Package.Counter++;

    //Encrypt the data
    Encrypt_Payload(Data_Tx, Session_Data->AppSKey, &Message);

    //Load Data
    for(i = 0; i < Data_Tx->Counter; i++)
    {
      RFM_Data[RFM_Package.Counter + i] = Data_Tx->Data[i];
    }

    //Add data Lenth to package counter
    RFM_Package.Counter = RFM_Package.Counter + Data_Tx->Counter;
  }

  //Calculate MIC
  Construct_Data_MIC(&RFM_Package, Session_Data, &Message);

  //Load MIC in package
  for(i = 0; i < 4; i++)
  {
    RFM_Data[i + RFM_Package.Counter] = Message.MIC[i];
  }

  //Add MIC length to RFM package length
  RFM_Package.Counter = RFM_Package.Counter + 4;

  //Send Package
  RFM_Send_Package(&RFM_Package, LoRa_Settings);

  //Raise Frame counter
  if(*Session_Data->Frame_Counter != 0xFFFF)
  {
    //Raise frame counter
    *Session_Data->Frame_Counter = *Session_Data->Frame_Counter + 1;
  }
  else
  {
    *Session_Data->Frame_Counter = 0x0000;
  }

  //Change channel for next message if hopping is activated
  if(LoRa_Settings->Channel_Hopping == 0x01)
  {
    if(LoRa_Settings->Channel_Tx < 0x07)
    {
      LoRa_Settings->Channel_Tx++;
    }
    else
    {
      LoRa_Settings->Channel_Tx = 0x00;
    }
  }
}


/*
*****************************************************************************************
* Description : Function that is used to receive a LoRaWAN message and retrieve the data from the RFM
*               Also checks on CRC, MIC and Device Address
*               This function is used for Class A and C motes.
*
* Arguments   : *Data_Rx pointer to receive buffer
*				*Session_Data pointer to sLoRa_Session sturct
*				*OTAA_Data pointer to sLoRa_OTAA struct
*				*Message_Rx pointer to sLoRa_Message struct used for the received message information
*				*LoRa_Settings pointer to sSetting struct
*****************************************************************************************
*/
void LORA_Receive_Data(sBuffer *Data_Rx, sLoRa_Session *Session_Data, sLoRa_OTAA *OTAA_Data, sLoRa_Message *Message, sSettings *LoRa_Settings)
{
	unsigned char i;

    //Initialise RFM buffer
	unsigned char RFM_Data[64];
	sBuffer RFM_Package = {&RFM_Data[0], 0x00};

	unsigned char MIC_Check;
  unsigned char Address_Check;

	unsigned char Frame_Options_Length;

	unsigned char Data_Location;

	message_t Message_Status = NO_MESSAGE;

	//If it is a type A device switch RFM to single receive
	if(LoRa_Settings->Mote_Class == 0x00)
	{
		Message_Status = RFM_Single_Receive(LoRa_Settings);
	}
	else
	{
		//Switch RFM to standby
		RFM_Switch_Mode(0x01);

		Message_Status = NEW_MESSAGE;
	}

	//If there is a message received get the data from the RFM
	if(Message_Status == NEW_MESSAGE)
	{
		Message_Status = RFM_Get_Package(&RFM_Package);

		//If mote class C switch RFM back to continuous receive
		if(LoRa_Settings->Mote_Class == 0x01)
		{
			//Switch RFM to Continuous Receive
			RFM_Continuous_Receive(LoRa_Settings);
		}
	}

	//if CRC ok breakdown package
	if(Message_Status == CRC_OK)
	{
		//Get MAC_Header
    	Message->MAC_Header = RFM_Data[0];

		//Join Accept message
		if(Message->MAC_Header == 0x20)
		{
			//Copy the data into the data array
			for(i = 0x00; i < RFM_Package.Counter; i++)
			{
				Data_Rx->Data[i] = RFM_Package.Data[i];
			}

			//Set data counter
			Data_Rx->Counter = RFM_Package.Counter;

			//Decrypt the data
			for(i = 0x00; i < ((Data_Rx->Counter - 1) / 16); i++)
			{
				AES_Encrypt(&(Data_Rx->Data[(i*16)+1]),OTAA_Data->AppKey);
			}

			//Calculate MIC
			//Remove MIC from number of bytes
			Data_Rx->Counter -= 4;

			//Get MIC
			Calculate_MIC(Data_Rx, OTAA_Data->AppKey, Message);

			//Clear MIC check counter
			MIC_Check = 0x00;

			//Compare MIC
			for(i = 0x00; i < 4; i++)
			{
				if(Data_Rx->Data[Data_Rx->Counter + i] == Message->MIC[i])
				{
					MIC_Check++;
				}
			}

			//Check if MIC compares
			if(MIC_Check == 0x04)
			{
				Message_Status = MIC_OK;

				Serial.write("MIC OK");
				UART_Send_Newline();
			}
			else
			{
				Message_Status = WRONG_MESSAGE;

				//Send NOK
				Serial.write("MIC NOK");
				UART_Send_Newline();
			}

			//Get Key's and data from package when MIC is OK
			if(Message_Status == MIC_OK)
			{
				//Get AppNonce
				OTAA_Data->AppNonce[0] = Data_Rx->Data[1];
				OTAA_Data->AppNonce[1] = Data_Rx->Data[2];
				OTAA_Data->AppNonce[2] = Data_Rx->Data[3];

				//Get Net ID
				OTAA_Data->NetID[0] = Data_Rx->Data[4];
				OTAA_Data->NetID[1] = Data_Rx->Data[5];
				OTAA_Data->NetID[2] = Data_Rx->Data[6];

				//Get session Device address
				Session_Data->DevAddr[3] = Data_Rx->Data[7];
				Session_Data->DevAddr[2] = Data_Rx->Data[8];
				Session_Data->DevAddr[1] = Data_Rx->Data[9];
				Session_Data->DevAddr[0] = Data_Rx->Data[10];

				//Calculate Network Session Key
				Session_Data->NwkSKey[0] = 0x01;

				//Load AppNonce
				Session_Data->NwkSKey[1] = OTAA_Data->AppNonce[0];
				Session_Data->NwkSKey[2] = OTAA_Data->AppNonce[1];
				Session_Data->NwkSKey[3] = OTAA_Data->AppNonce[2];

				//Load NetID
				Session_Data->NwkSKey[4] = OTAA_Data->NetID[0];
				Session_Data->NwkSKey[5] = OTAA_Data->NetID[1];
				Session_Data->NwkSKey[6] = OTAA_Data->NetID[2];

				//Load Dev Nonce
				Session_Data->NwkSKey[7] = OTAA_Data->DevNonce[0];
				Session_Data->NwkSKey[8] = OTAA_Data->DevNonce[1];

				//Pad with zeros
				Session_Data->NwkSKey[9] = 0x00;
				Session_Data->NwkSKey[10] = 0x00;
				Session_Data->NwkSKey[11] = 0x00;
				Session_Data->NwkSKey[12] = 0x00;
				Session_Data->NwkSKey[13] = 0x00;
				Session_Data->NwkSKey[14] = 0x00;
				Session_Data->NwkSKey[15] = 0x00;

				//Copy to AppSkey
				for(i = 0x00; i < 16; i++)
				{
					Session_Data->AppSKey[i] = Session_Data->NwkSKey[i];
				}

				//Change first byte of AppSKey
				Session_Data->AppSKey[0] = 0x02;

				//Calculate the keys
				AES_Encrypt(Session_Data->NwkSKey,OTAA_Data->AppKey);
				AES_Encrypt(Session_Data->AppSKey,OTAA_Data->AppKey);

				//Reset Frame counter
				*Session_Data->Frame_Counter = 0x0000;

				//Send dat over USB
				Serial.write("DevAddr: ");
				UART_Send_Data(Session_Data->DevAddr,0x04);
				UART_Send_Newline();

				Serial.write("NwkSKey: ");
				UART_Send_Data(Session_Data->NwkSKey,0x10);
				UART_Send_Newline();

				Serial.write("AppSKey: ");
				UART_Send_Data(Session_Data->AppSKey,0x10);
				UART_Send_Newline();
				UART_Send_Newline();

				//Clear Data counter
				Data_Rx->Counter = 0x00;
			}
		}

		//Data message
		if(Message->MAC_Header == 0x40 || Message->MAC_Header == 0x60 || Message->MAC_Header == 0x80 || Message->MAC_Header == 0xA0)
		{
			//Get device address from received data
			Message->DevAddr[0] = RFM_Data[4];
			Message->DevAddr[1] = RFM_Data[3];
			Message->DevAddr[2] = RFM_Data[2];
			Message->DevAddr[3] = RFM_Data[1];

			//Get frame control field
			Message->Frame_Control = RFM_Data[5];

			 //Get frame counter
			Message->Frame_Counter = RFM_Data[7];
			Message->Frame_Counter = (Message->Frame_Counter << 8) + RFM_Data[6];

			//Lower Package length with 4 to remove MIC length
			RFM_Package.Counter -= 4;

			//Calculate MIC
			Construct_Data_MIC(&RFM_Package, Session_Data, Message);

			MIC_Check = 0x00;

            //Compare MIC
			for(i = 0x00; i < 4; i++)
			{
				if(RFM_Data[RFM_Package.Counter + i] == Message->MIC[i])
				{
					MIC_Check++;
				}
			}

			//Check MIC
      		if(MIC_Check == 0x04)
      		{
      		  Message_Status = MIC_OK;

      		  Serial.write("MIC OK");
      		}
      		else
      		{
      		  Message_Status = WRONG_MESSAGE;

      		  //Send NOK
      		  Serial.write("MIC NOK");

      		}

      		UART_Send_Newline();

      		Address_Check = 0;

      		//Check address
      		if(MIC_Check == 0x04)
      		{
			      for(i = 0x00; i < 4; i++)
			      {
			        if(Session_Data->DevAddr[i] == Message->DevAddr[i])
			        {
				        Address_Check++;
			        }
			      }
      		}

		  	if(Address_Check == 0x04)
		  	{
				Message_Status = ADDRESS_OK;

				Serial.write("ADDRESS OK");
		  	}
		  	else
		  	{
				Message_Status = WRONG_MESSAGE;

				//Send NOK
				Serial.write("ADDRESS NOK");

		  	}

		  	UART_Send_Newline();

			//if the address is OK then decrypt the data
			//Send the data to USB
			if(Message_Status == ADDRESS_OK)
			{
				//Send Mac Header
				Serial.write("Mac Header: ");
				UART_Send_Data(&Message->MAC_Header, 0x01);
				UART_Send_Newline();

				//Send Dev addr
				Serial.write("Dev addr: ");
				UART_Send_Data(Message->DevAddr,0x04);
				UART_Send_Newline();

				//Send Frame control field
				Serial.write("Frame Control: ");
				UART_Send_Data(&Message->Frame_Control,0x01);
				UART_Send_Newline();

				//Send frame counter
				Serial.write("Frame Counter: ");
				UART_Send_Data(&RFM_Data[7],0x01);
				UART_Send_Data(&RFM_Data[6],0x01);
				UART_Send_Newline();

				Data_Location = 8;

				//Get length of frame options field
				Frame_Options_Length = (Message->Frame_Control & 0x0F);

				//Add length of frame options field to data location
				Data_Location = Data_Location + Frame_Options_Length;

				//Check if ther is data in the package
				if(RFM_Package.Counter == Data_Location)
				{
					Data_Rx->Counter = 0x00;
				}
				else
				{
					//Get port field when ther is data
					Message->Frame_Port = RFM_Data[8];

					//Send Port field
					Serial.write("Frame Port: ");
					UART_Send_Data(&(Message->Frame_Port),0x01);
					UART_Send_Newline();

					//Calculate the amount of data in the package
					Data_Rx->Counter = (RFM_Package.Counter - Data_Location -1);

					//Correct the data location by 1 for the Fport field
					Data_Location = (Data_Location + 1);
				}

				//Copy and decrypt the data
				if(Data_Rx->Counter != 0x00)
				{
					for(i = 0; i < Data_Rx->Counter; i++)
					{
						Data_Rx->Data[i] = RFM_Data[Data_Location + i];
					}

         //Check frame port fiels. When zero it is a mac command message encrypted with NwkSKey
         if(Message->Frame_Port == 0x00)
         {
          Encrypt_Payload(Data_Rx, Session_Data->NwkSKey, Message);
         }
         else
         {
          Encrypt_Payload(Data_Rx, Session_Data->AppSKey, Message);
         }

					Message_Status = MESSAGE_DONE;
				}
			}
		}

		if(Message_Status == WRONG_MESSAGE)
		{
			Data_Rx->Counter = 0x00;
 		}
	}
}

/*
*****************************************************************************************
* Description : Function that is used to send a join request to a network.
*
* Arguments   : *OTAA_Data pointer to sLoRa_OTAA struct
*				*LoRa_Settings pointer to sSetting struct
*****************************************************************************************
*/
void LoRa_Send_JoinReq(sLoRa_OTAA *OTAA_Data, sSettings *LoRa_Settings)
{
    unsigned char i;

	//Initialise RFM data buffer
    unsigned char RFM_Data[23];
    sBuffer RFM_Package = { &RFM_Data[0], 0x00};

 	//Initialise message sturct
    sLoRa_Message Message;

    Message.MAC_Header = 0x00; //Join request
    Message.Direction = 0x00; //Set up Direction

    //Construct OTAA Request message
    //Load Header in package
    RFM_Data[0] = Message.MAC_Header;

    //Load AppEUI in package
    for(i = 0x00; i < 8; i++)
    {
        RFM_Data[i+1] = OTAA_Data->AppEUI[7-i];
    }

    //Load DevEUI in package
    for(i= 0x00; i < 8; i++)
    {
        RFM_Data[i+9] = OTAA_Data->DevEUI[7-i];
    }

    //Generate DevNonce
    Generate_DevNonce(OTAA_Data->DevNonce);

    //Load DevNonce in package
    RFM_Data[17] = OTAA_Data->DevNonce[0];
    RFM_Data[18] = OTAA_Data->DevNonce[1];

    //Set length of package
    RFM_Package.Counter = 19;

    //Get MIC
    Calculate_MIC(&RFM_Package, OTAA_Data->AppKey, &Message);

    //Load MIC in package
    RFM_Data[19] = Message.MIC[0];
    RFM_Data[20] = Message.MIC[1];
    RFM_Data[21] = Message.MIC[2];
    RFM_Data[22] = Message.MIC[3];

    //Set lenght of package to the right length
    RFM_Package.Counter = 23;

    //Send Package
    RFM_Send_Package(&RFM_Package, LoRa_Settings);
}

/*
*****************************************************************************************
* Description : Function that is used to generate device nonce used in the join reqeust function
*				This is based on a psuedo random function in the arduino library
*
* Arguments   : *Devnonce pointer to the devnonce arry of withc is unsigned char[2]
*****************************************************************************************
*/
void Generate_DevNonce(unsigned char *DevNonce)
{
  unsigned int RandNumber;

  RandNumber = random(0xFFFF);

  DevNonce[0] = RandNumber & 0x00FF;
  DevNonce[1] = (RandNumber >> 8) & 0x00FF;
}
